File: 19Z310 - Data Structures Laboratory/Final Report/Kth Smallest Element/timerunner.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

// The provided function kth and kthSmallest

int kthsmallest(struct TreeNode *rootnode, int * counterpointer, int k)
{
    if (rootnode == NULL || *counterpointer >= k) {
        return -1;
    }

    int left = kthsmallest(rootnode->left, counterpointer, k);

    if (*counterpointer == k) {
        return left; // return the kth smallest from the left subtree if found
    }

    *counterpointer = *counterpointer + 1;

    if (*counterpointer == k) {
        return rootnode->val;
    }

    return kthsmallest(rootnode->right, counterpointer, k);
}

struct TreeNode* newNode(int val) {
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = NULL;
    node->right = NULL;
    return node;
}

// Function to insert nodes into the tree
struct TreeNode* insert(struct TreeNode* root, int val) {
    if (root == NULL) return newNode(val);

    if (val < root->val)
        root->left = insert(root->left, val);
    else if (val > root->val)
        root->right = insert(root->right, val);

    return root;
}

// Function to generate a BST with given elements
struct TreeNode* generateBST(int arr[], int n) {
    struct TreeNode* root = NULL;
    for (int i = 0; i < n; i++) {
        root = insert(root, arr[i]);
    }
    return root;
}

// Function to measure execution time of kthSmallest function
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

double measureTime(struct TreeNode* root, int k, int *position) {
    struct timespec start, end;
    double cpu_time_used;
    int counter = 0;

    clock_gettime(CLOCK_MONOTONIC, &start);
    *position = kthsmallest(root, &counter, k); // Corrected function name
    clock_gettime(CLOCK_MONOTONIC, &end);

    cpu_time_used = (end.tv_sec - start.tv_sec) + (double)(end.tv_nsec - start.tv_nsec) / 1e9;
    return cpu_time_used;
}

int main() {
    srand(time(NULL)); // Seed for random numbers
    FILE *fp;

    // Open the file in write mode
    fp = fopen("benchmarks.csv", "w");
    if (fp == NULL) {
        printf("Error opening file.\n");
        return 1;
    }

    // Write header to the CSV file
    fprintf(fp, "Test Case,Array Size,Kth Smallest Element Position,Runtime\n");

    for (int i = 0; i < 100000; i++) {
        int size = rand() % 100000 + 1; // Generate array size between 1 to 100000
        int testData[size];

        // Generate random integers for test data
        for (int j = 0; j < size; j++) {
            testData[j] = rand() % 1000; // Random integers between 0 to 999
        }

        struct TreeNode* root = generateBST(testData, size);

        int position;
        double runtime = measureTime(root, rand() % size + 1, &position); // Random k value between 1 to size
        if(position!=0){
            fprintf(fp, "%d,%d,%d,%f\n", i + 1, size, position, runtime);
        }
    }

    fclose(fp); // Close the file

    return 0;
}

File: 19Z310 - Data Structures Laboratory/Final Report/Serialize-Deserialize/timerunner.py
import csv
import random
import time
from collections import deque

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:
    def serialize(self, root):
        if not root:
            return "[]"
        
        queue = deque([root])
        result = []
        
        while queue:
            node = queue.popleft()
            if node:
                result.append(str(node.val))
                queue.append(node.left)
                queue.append(node.right)
            else:
                result.append('null')
        
        return '[' + ','.join(result) + ']'

    def deserialize(self, data):
        if data == "[]":
            return None
        
        nodes, i = data[1:-1].split(','), 1
        root = TreeNode(int(nodes[0]))
        queue = deque([root])
        
        while queue:
            node = queue.popleft()
            if nodes[i] != 'null':
                node.left = TreeNode(int(nodes[i]))
                queue.append(node.left)
            i += 1
            if nodes[i] != 'null':
                node.right = TreeNode(int(nodes[i]))
                queue.append(node.right)
            i += 1
        
        return root

def generate_random_tree(depth):
    if depth > 0 and random.random() < 0.9:  # 90% chance to continue to generate a subtree
        node = TreeNode(random.randint(0, 100))
        node.left = generate_random_tree(depth - 1)
        node.right = generate_random_tree(depth - 1)
        return node
    return None  # 10% chance to return a leaf node (None)

def main():
    codec = Codec()
    with open('benchmarks.csv', 'w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(["S.NO", "Testcase Size", "Runtime"])
        for i in range(1, 10001):  # Generate 10000 test cases
            tree = generate_random_tree(5)  # Generate a random tree with depth up to 5
            start_time = time.time()
            serialized = codec.serialize(tree)
            deserialized = codec.deserialize(serialized)
            end_time = time.time()
            writer.writerow([i, len(serialized), end_time - start_time])

if __name__ == "__main__":
    main()